cpp/src/hash/sha_hash.cu

```diff
@@ -0,0 +1,630 @@
+{
+  return !is_chrono(dt) && (is_fixed_width(dt) || (dt.id() == type_id::STRING));
+}
+
+CUDA_DEVICE_CALLABLE uint32_t rotate_bits_left(uint32_t x, int8_t r)
+{
+  // This function is equivalent to (x << r) | (x >> (32 - r))
+  return __funnelshift_l(x, x, r);
+}
+
+CUDA_DEVICE_CALLABLE uint32_t rotate_bits_right(uint32_t x, int8_t r)
+{
+  // This function is equivalent to (x >> r) | (x << (32 - r))
+  return __funnelshift_r(x, x, r);
+}
+
+CUDA_DEVICE_CALLABLE uint64_t rotate_bits_right(uint64_t x, int8_t r)
+{
+  return (x >> r) | (x << (64 - r));
+}
+
+// Swap the endianness of a 32 bit value
+CUDA_DEVICE_CALLABLE uint32_t swap_endian(uint32_t x)
+{
+  // The selector 0x0123 reverses the byte order
+  return __byte_perm(x, 0, 0x0123);
+}
+
+// Swap the endianness of a 64 bit value
+// There is no CUDA intrinsic for permuting bytes in 64 bit integers
+CUDA_DEVICE_CALLABLE uint64_t swap_endian(uint64_t x)
+{
+  // Reverse the endianness of each 32 bit section
+  uint32_t low_bits  = swap_endian(static_cast<uint32_t>(x));
+  uint32_t high_bits = swap_endian(static_cast<uint32_t>(x >> 32));
+  // Reassemble a 64 bit result, swapping the low bits and high bits
+  return (static_cast<uint64_t>(low_bits) << 32) | (static_cast<uint64_t>(high_bits));
+};
+
+}  // namespace
+
+template <typename HasherT>
+struct SHAHash {
+  /**
+   * @brief Execute SHA on input data chunks.
+   *
+   * This accepts arbitrary data, handles it as bytes, and calls the hash step
+   * when the buffer is filled up to message_chunk_size bytes.
+   */
+  template <typename TKey, typename Hasher = HasherT>
+  void CUDA_DEVICE_CALLABLE process(TKey const& key,
+                                    typename Hasher::sha_intermediate_data* hash_state) const
+  {
+    uint32_t const len  = sizeof(TKey);
+    uint8_t const* data = reinterpret_cast<uint8_t const*>(&key);
+    hash_state->message_length += len;
+
+    if (hash_state->buffer_length + len < Hasher::message_chunk_size) {
+      std::memcpy(hash_state->buffer + hash_state->buffer_length, data, len);
+      hash_state->buffer_length += len;
+    } else {
+      uint32_t copylen = Hasher::message_chunk_size - hash_state->buffer_length;
+
+      std::memcpy(hash_state->buffer + hash_state->buffer_length, data, copylen);
+      Hasher::hash_step(hash_state);
+
+      while (len > Hasher::message_chunk_size + copylen) {
+        std::memcpy(hash_state->buffer, data + copylen, Hasher::message_chunk_size);
+        Hasher::hash_step(hash_state);
+        copylen += Hasher::message_chunk_size;
+      }
+
+      std::memcpy(hash_state->buffer, data + copylen, len - copylen);
+      hash_state->buffer_length = len - copylen;
+    }
+  }
+
+  /**
+   * @brief Finalize SHA element processing.
+   *
+   * This method fills the remainder of the message buffer with zeros, appends
+   * the message length (in another step of the hash, if needed), and performs
+   * the final hash step.
+   */
+  template <typename Hasher = HasherT>
```

Comment:
Why is this a template?

Reply:
@jrhemstad I was facing a problem with accepting template-dependent typenames as arguments to methods in the CRTP base class.

Minimal reproduction of problem: https://godbolt.org/z/PcGj6hjaK

The issue is also similar to the one described here: https://www.py4u.net/discuss/95594

Any insights on a cleaner pattern would be welcome.

Reply:
Aha, understood. It's the same issue as when trying to SFINAE a member function of a class template.